## Problem Statement & Context

Customer churn is a major challenge for companies that rely on subscription-based services such as Internet Service Providers (ISPs). When customers leave, companies not only lose recurring revenue but also incur additional costs to acquire new users.

This project focuses on a binary classification task:

Predict whether a customer is likely to churn (True) or remain (False).

Customer churn affects multiple stakeholders. Customers may experience declining service quality or unmet expectations before leaving, while customer success and marketing teams need to identify at-risk users early enough to intervene. For businesses, high churn rates directly impact profitability and long-term growth.

A reliable churn prediction model can support better decision-making by enabling teams to proactively target customers who are likely to leave. This includes offering personalized incentives, improving support interactions, and optimizing retention strategies. By focusing efforts on the most at-risk customers, companies can allocate resources more efficiently and improve overall customer experience.

## Hypothesis & Intended Impact

This project is based on the assumption that customer behavior, usage patterns, and account-related information contain enough signal to predict churn accurately. By analyzing these features, the model should be able to identify customers who are at risk of leaving before the actual event occurs.

If the model performs well, it can enable more proactive and targeted retention strategies. Instead of applying broad and often inefficient campaigns, companies can focus on customers who are most likely to churn. This can lead to improved customer satisfaction, as users receive more relevant support and offers tailored to their needs.

From a business perspective, better churn prediction can reduce customer acquisition costs, increase customer lifetime value, and improve overall operational efficiency. More broadly, focusing on retaining existing customers rather than constantly replacing them supports more sustainable and resource-efficient business practices.

## Dataset Discovery & Selection

To support this classification task, three datasets were initially considered: a telecom dataset, a bank customer churn dataset, and an ISP dataset. Each dataset was evaluated based on alignment with the problem, data quality, scale, documentation, and ethical considerations.

The ISP dataset was ultimately selected because it provides a strong balance across these dimensions. It is well-documented, sufficiently large, and directly aligned with the problem of predicting customer churn in a subscription-based service context. In addition, the dataset is released under a public domain (CC0) license, which removes potential legal or usage constraints.

Compared to the alternatives, the ISP dataset offers high-quality features, reliable labels, and minimal access barriers. While the bank dataset also performed well in evaluation, the ISP dataset was chosen due to its closer relevance to the specific churn scenario being modeled.

## Dataset Justification & Target Variable

The ISP dataset is well-suited for this problem because it includes detailed information about customer behavior, service usage, and account history. These types of features are commonly associated with churn and provide a strong foundation for building predictive models.

The target variable in this dataset is Churn, which indicates whether a customer has left the service. It is defined as a binary variable:

True: the customer has churned
False: the customer has remained

This target is directly relevant to the business problem and is both meaningful and actionable. Predicting churn allows stakeholders to intervene before the customer leaves, making it possible to influence outcomes rather than simply observe them.

## Features Overview

The dataset contains a mix of numerical, categorical, and behavioral features that describe different aspects of customer activity.

Numerical features include variables such as usage levels and billing amounts, which can indicate how actively a customer is using the service. Categorical features describe subscription types or service configurations, which may influence customer satisfaction. Behavioral features capture patterns over time, such as changes in usage or interaction frequency. Account-related features, such as tenure, provide context about the customer’s relationship with the service.

Several features are expected to be particularly informative. For example, customers with shorter tenure may be more likely to churn, while unusual changes in usage or billing behavior may signal dissatisfaction. Overall, the feature set appears rich enough to support accurate predictions.

## Data Overview

The dataset contains approximately 72,000 customer records, making it large enough to support robust model training and evaluation.

The target variable is reasonably well distributed:

- Around 40,050 customers have churned
- Around 32,224 customers have not churned

This indicates a moderate class balance, which is beneficial for classification tasks. Since neither class is extremely underrepresented, the model can learn patterns for both churners and non-churners without requiring heavy rebalancing techniques.

## Ethical Considerations

Ethical considerations were taken into account during the data preparation stage. One feature, gender, was identified as a sensitive attribute and removed from the dataset. This decision was made to reduce the risk of introducing bias into the model and to avoid predictions being directly influenced by protected characteristics.

Even after removing explicit sensitive features, there is still a possibility of indirect bias. Some variables may act as proxies for demographic information, which could lead to unfair outcomes for certain groups. This highlights the importance of evaluating fairness in later stages of the project.

From a regulatory perspective, this system can be considered a limited-risk application, as it supports decision-making but does not directly automate high-stakes outcomes. However, careful monitoring is still necessary to ensure responsible use.

## Real-World Framing

The prediction task can be clearly defined in a real-world context.

The unit of analysis is an individual customer, and the event of interest is whether that customer churns. Predictions are made with a horizon of 30 days, meaning the model estimates the likelihood of churn within the next month. These predictions can be generated on a daily basis using the most recent available data.

The outputs of the model are intended to support actions such as targeted retention campaigns, personalized offers, or proactive customer support. The main stakeholders include customer success teams, marketing teams, and the customers themselves.

There is an important trade-off between recall and precision. In this context, recall is especially important, as failing to identify a customer who is about to churn can result in lost revenue. However, improving recall may lead to more false positives, which can increase the cost of retention efforts.

## Success Metrics

The primary evaluation metric for this project is recall for the churn class. This reflects the importance of identifying as many at-risk customers as possible.

Additional metrics such as precision and F1-score will also be considered to provide a more balanced view of model performance.

Beyond model-level metrics, success can also be measured using business outcomes. These include a reduction in churn rate, improved effectiveness of retention campaigns, and an increase in customer lifetime value. Together, these metrics provide a more complete picture of the model’s real-world impact.

## Key dataset characteristics
|Metric| Value|
|----------|-------------|
|Number of samples|     ~72000       |
|Number of classes|           2  |
|Class imbalance|              Moderate|
|Features|              Behavioral, account, and usage data      |
|Target variable|            Churn       |



## Data preprocessing

In [209]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [210]:
df_isp = pd.read_csv(r"..\Data\Interim\internet_service_churn.csv")
df_isp = df_isp.rename(columns={
  'reamining_contract': 'remaining_contract_y',
  'id': 'customer_id',
  'is_tv_subscriber': 'tv_subscriber',
  'is_movie_package_subscriber': 'movie_package_subscriber',
  'subscription_age': 'subscription_age_y',
  'bill_avg': 'avg_bill_3m',
  'download_avg': 'avg_download_usage_3m',
  'upload_avg': 'avg_upload_usage_3m',
  'download_over_limit': 'exceeded_download_limit_3m'
  })
display(df_isp.head())
display(df_isp.shape)
display(df_isp.describe())
df_isp.info()

,customer_id,tv_subscriber,movie_package_subscriber,subscription_age_y,avg_bill_3m,remaining_contract_y,service_failure_count,avg_download_usage_3m,avg_upload_usage_3m,exceeded_download_limit_3m,churn
0,15,1,0,11.95,25,0.14,0,8.4,2.3,0,0
1,18,0,0,8.22,0,NaN,0,0.0,0.0,0,1
2,23,1,0,8.91,16,0.00,0,13.7,0.9,0,1
3,27,0,0,6.87,21,NaN,1,0.0,0.0,0,1
4,34,0,0,6.39,0,NaN,0,0.0,0.0,0,1


(72274, 11)

,customer_id,tv_subscriber,movie_package_subscriber,subscription_age_y,avg_bill_3m,remaining_contract_y,service_failure_count,avg_download_usage_3m,avg_upload_usage_3m,exceeded_download_limit_3m,churn
count,7.227400e+04,72274.000000,72274.000000,72274.000000,72274.000000,50702.000000,72274.000000,71893.000000,71893.000000,72274.000000,72274.000000
mean,8.463182e+05,0.815259,0.334629,2.450051,18.942483,0.716039,0.274234,43.689911,4.192076,0.207613,0.554141
std,4.891022e+05,0.388090,0.471864,2.034990,13.215386,0.697102,0.816621,63.405963,9.818896,0.997123,0.497064
min,1.500000e+01,0.000000,0.000000,-0.020000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,4.222165e+05,1.000000,0.000000,0.930000,13.000000,0.000000,0.000000,6.700000,0.500000,0.000000,0.000000
50%,8.477840e+05,1.000000,0.000000,1.980000,19.000000,0.570000,0.000000,27.800000,2.100000,0.000000,1.000000
75%,1.269562e+06,1.000000,1.000000,3.300000,22.000000,1.310000,0.000000,60.500000,4.800000,0.000000,1.000000
max,1.689744e+06,1.000000,1.000000,12.800000,406.000000,2.920000,19.000000,4415.200000,453.300000,7.000000,1.000000


<class 'pandas.DataFrame'>
RangeIndex: 72274 entries, 0 to 72273
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   customer_id                 72274 non-null  int64  
 1   tv_subscriber               72274 non-null  int64  
 2   movie_package_subscriber    72274 non-null  int64  
 3   subscription_age_y          72274 non-null  float64
 4   avg_bill_3m                 72274 non-null  int64  
 5   remaining_contract_y        50702 non-null  float64
 6   service_failure_count       72274 non-null  int64  
 7   avg_download_usage_3m       71893 non-null  float64
 8   avg_upload_usage_3m         71893 non-null  float64
 9   exceeded_download_limit_3m  72274 non-null  int64  
 10  churn                       72274 non-null  int64  
dtypes: float64(4), int64(7)
memory usage: 6.1 MB


## Functions Used

In [211]:
# Define detect_variable_types function
def detect_variable_types(df, target_col=None):
    """Automatically detect variable types in a DataFrame"""
    types = {}
    
    for col in df.columns:
        if col == target_col:
            types[col] = 'target'
        elif pd.api.types.is_datetime64_any_dtype(df[col]):
            types[col] = 'datetime'
        elif pd.api.types.is_numeric_dtype(df[col]):
            if df[col].nunique() == 2:
                types[col] = 'binary'
            else:
                types[col] = 'numerical'
        elif df[col].nunique() < 20:  # Threshold for categorical*
            types[col] = 'categorical'
        elif df[col].dtype == 'object':
            # Check if it's text (long strings) or categorical*
            avg_length = df[col].astype(str).str.len().mean()
            if avg_length > 50:
                types[col] = 'text'
            else:
                types[col] = 'categorical'
    
    return types


In [212]:
# Check missing values
def missing_values(df):
    missing_v = df.isnull().sum()
    missing_percentage = (missing_v / len(df)) * 100
    zero_v = (df.loc[:, df.nunique() > 2] == 0).sum()
    zero_v_percentage = (zero_v/len(df))*100

    summary = pd.DataFrame({
        "Missing values": missing_v,
        "Missing %": missing_percentage,
        "Zero values": zero_v,
        "Zero %": zero_v_percentage
    })

    print("Missing value summary:\n")

    return summary

In [213]:
# Check range and distribution
# Quick sanity checks for numerical data
def basic_range_check(df):
    numeric_cols = df.select_dtypes(include=['number']).columns
    
    for col in numeric_cols:
        print(f"{col}: min={df[col].min()}, max={df[col].max()}")
        
        # Flag obviously problematic values
        if col.lower() in ['subscription_age'] and (df[col].min() < 0 or df[col].max() > 150):
            print(f"  ⚠️ Suspicious subscription_age values detected")
        elif col.lower() in ['bill_avg', 'download_avg', 'upload_avg', 'download_over_limit', 'remaining_contract'] and df[col].min() < 0:
            print(f"  ⚠️ Negative financial values detected")
    

In [214]:
# Check duplicates
def duplicates(df, key=None):
  
  exact_dup = df.duplicated().sum()
  print(f"exact duplicates: {exact_dup}")

  key_dup = None
  if key is not None:
    key_dup = df.duplicated(subset=key).sum()
    print(f"key_duplicates: {key_dup}")
  
  return (exact_dup, key_dup)

## Preprocessing

### Detect variable types

In [215]:
# Call detect_variable_types function
result = detect_variable_types(df_isp)
print(result)

{'customer_id': 'numerical', 'tv_subscriber': 'binary', 'movie_package_subscriber': 'binary', 'subscription_age_y': 'numerical', 'avg_bill_3m': 'numerical', 'remaining_contract_y': 'numerical', 'service_failure_count': 'numerical', 'avg_download_usage_3m': 'numerical', 'avg_upload_usage_3m': 'numerical', 'exceeded_download_limit_3m': 'numerical', 'churn': 'binary'}


In [216]:
# Create df_col_types
df_col_types = pd.Series(result).reset_index(name="Type")
df_col_types.columns = ["Column", "Type"]
df_col_types

,Column,Type
0,customer_id,numerical
1,tv_subscriber,binary
2,movie_package_subscriber,binary
3,subscription_age_y,numerical
4,avg_bill_3m,numerical
5,remaining_contract_y,numerical
6,service_failure_count,numerical
7,avg_download_usage_3m,numerical
8,avg_upload_usage_3m,numerical
9,exceeded_download_limit_3m,numerical


### Check missing values

In [217]:
# Call missing_values fuction
missing_summary = missing_values(df_isp)
missing_summary

Missing value summary:



,Missing values,Missing %,Zero values,Zero %
avg_bill_3m,0,0.000000,5173.0,7.157484
avg_download_usage_3m,381,0.527161,10941.0,15.138224
avg_upload_usage_3m,381,0.527161,11674.0,16.152420
churn,0,0.000000,NaN,NaN
customer_id,0,0.000000,0.0,0.000000
exceeded_download_limit_3m,0,0.000000,68373.0,94.602485
movie_package_subscriber,0,0.000000,NaN,NaN
remaining_contract_y,21572,29.847525,16363.0,22.640230
service_failure_count,0,0.000000,60407.0,83.580541
subscription_age_y,0,0.000000,278.0,0.384647


In [218]:
# Check reason for missing data in remaining_contract_y
missing_remaining = ((df_isp['remaining_contract_y'].isna()) & (df_isp['churn'] == 1)).sum()
print(missing_remaining)

19719


**Missing Value Analysis**<br>
The column `remaining_contract_y` initially contains 21,572 missing values (29.85%). However, further investigation shows that 19,719 of these missing values correspond to customers who have churned (`churn = 1`).

This pattern is logically consistent in a real-world context, since customers who have churned are no longer under an active contract. Therefore, the absence of a remaining contract value for these customers is expected and informative, rather than a data quality issue.

After excluding these churn-related cases, the number of true missing values reduces to 1,853, which represents only 2.56% of the total dataset.

### Check the range of variables

In [219]:
# Call basic_range function
range_summary = basic_range_check(df_isp)
range_summary

customer_id: min=15, max=1689744
tv_subscriber: min=0, max=1
movie_package_subscriber: min=0, max=1
subscription_age_y: min=-0.02, max=12.8
avg_bill_3m: min=0, max=406
remaining_contract_y: min=0.0, max=2.92
service_failure_count: min=0, max=19
avg_download_usage_3m: min=0.0, max=4415.2
avg_upload_usage_3m: min=0.0, max=453.3
exceeded_download_limit_3m: min=0, max=7
churn: min=0, max=1


In [220]:
# Further investigate on invalid values


### Check duplicates

In [221]:
df_isp.head()

,customer_id,tv_subscriber,movie_package_subscriber,subscription_age_y,avg_bill_3m,remaining_contract_y,service_failure_count,avg_download_usage_3m,avg_upload_usage_3m,exceeded_download_limit_3m,churn
0,15,1,0,11.95,25,0.14,0,8.4,2.3,0,0
1,18,0,0,8.22,0,NaN,0,0.0,0.0,0,1
2,23,1,0,8.91,16,0.00,0,13.7,0.9,0,1
3,27,0,0,6.87,21,NaN,1,0.0,0.0,0,1
4,34,0,0,6.39,0,NaN,0,0.0,0.0,0,1


In [222]:
# Call duplicates function
dups = duplicates(df_isp, key=('customer_id', 'subscription_age_y'))


exact duplicates: 0
key_duplicates: 0


## Data Quality Scorecard
| Columns                      | Missing Values (%) | Duplicate Records (%)  | Invalid Values (%) | Outliers (%)|
|------------------------------|----------------|---------------------|---------------|---------|
|  id                           | 0           | 0             | 5                    |           |
| is_tv_subscriber              | 0           | 0              | 5                    |          |
| is_movie_package_subscriber  | 0              | 0            | 5                     |        |
| subscription_age              | 0           | 0               | 3                    |        |
| bill_avg                      | 0            | 0             | 5                     |        |
| reamining_contract             | 2.56           | 0             | 5                     |        |
| service_failure_count        | 0             | 0              | 5                     |         |
| download_avg                  | 0.53             | 0              | 5                    |          |
| upload_avg                     | 0.53            | 0            | 38                  |           |
| download_over_limit            | 0             | 0             | 5                   |         |
| churn                           | 0         | 0             | 38                   |          |